OSDOCS-19890 created zero-trust assembly/modules for multi-cluster (#112274)
openshift/openshift-docs: modules/zero-trust-manager-deploy-istiocni.adoc

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-istiocni_{context}"]

> = Deploy {SMProductName} CNI on both clusters

Reviewer: gerund

[role="_abstract"]
Deploy the `IstioCNI` custom resource and create federated `ClusterSPIFFEID` resources on Cluster A and Cluster B. `IstioCNI` configures pod networking for {SMProductName} so sidecars can intercept traffic without privileged init containers. Federated `ClusterSPIFFEID` resources ensure workloads and gateways receive trust bundles from both clusters.

.Prerequisites

> * You have completed the "Configuring {SMProductName} for multi-cluster {spire-full} integration" steps.

Reviewer: similar comment. Action + for more info

> * The environment variables from the deploy SPIRE procedure are still set in your shell session.

Reviewer: Same as in East West procedure

* You have installed {SMProductName} 3 on both clusters.

.Procedure

> . Deploy `IstioCNI` on both clusters by running the following commands:

Reviewer: this is treated as a single command
```suggestion
. Deploy `IstioCNI` on both clusters by running the following command
:
```

+
[source,terminal,subs="+quotes,verbatim"]
----
$ for KUBECONFIG in "${CLUSTER_A_KUBECONFIG}" "${CLUSTER_B_KUBECONFIG}"; do
  oc new-project "${OSSM_CNI}" --kubeconfig="${KUBECONFIG}" 2>/dev/null || true

  oc apply --kubeconfig="${KUBECONFIG}" -f - <<EOF
apiVersion: sailoperator.io/v1
kind: IstioCNI
metadata:
  name: default
spec:
  namespace: ${OSSM_CNI}
EOF

 until oc get daemonset/istio-cni-node --kubeconfig="${KUBECONFIG}" -n "${OSSM_CNI}" &> /dev/null; do
  sleep 3
 done
  oc rollout status daemonset/istio-cni-node --kubeconfig="${KUBECONFIG}" -n "${OSSM_CNI}" --timeout=300s
done
----

> . Create the federated `ClusterSPIFFEID` resources on Cluster A:

Reviewer: ```suggestion
. Create the federated `ClusterSPIFFEID` resources on Cluster A by running the following command:
```

+
[NOTE]
====
Do not patch the default `ClusterSPIFFEID` (`zero-trust-workload-identity-manager-spire-default`). The {zero-trust-full} reconciles and reverts the manual changes. Instead, create custom `ClusterSPIFFEID` resources for the specific namespaces.
====
+
[source,terminal]
----
$ oc apply --kubeconfig="${CLUSTER_A_KUBECONFIG}" -f - <<EOF
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterSPIFFEID
metadata:
  name: sample-federation
spec:
  className: zero-trust-workload-identity-manager-spire
  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
  namespaceSelector:
    matchLabels:
      kubernetes.io/metadata.name: sample
  federatesWith:
    - "${CLUSTER_B_TRUST_DOMAIN}"
---
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterSPIFFEID
metadata:
  name: istio-system-federation
spec:
  className: zero-trust-workload-identity-manager-spire
  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
  namespaceSelector:
    matchLabels:
      kubernetes.io/metadata.name: istio-system
  federatesWith:
    - "${CLUSTER_B_TRUST_DOMAIN}"
EOF
----

> . Create federated `ClusterSPIFFEID` resources on Cluster B:

Reviewer: ```suggestion
. Create federated `ClusterSPIFFEID` resources on Cluster B by running the following command:
```

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-istiocni_{context}"]

> = Deploy {SMProductName} CNI on both clusters

Reviewer 2: ```suggestion
= Deploying the {SMProductName} CNI on both clusters
```

[role="_abstract"]

> Deploy the `IstioCNI` custom resource and create federated `ClusterSPIFFEID` resources on Cluster A and Cluster B. `IstioCNI` configures pod networking for {SMProductName} so sidecars can intercept traffic without privileged init containers. Federated `ClusterSPIFFEID` resources ensure workloads and gateways receive trust bundles from both clusters.

Reviewer 2: ```suggestion
Deploy the `IstioCNI` custom resource (CR) and create federated `ClusterSPIFFEID` resources on Cluster A and Cluster B. The `IstioCNI` CR configures pod networking for {SMProductName} so sidecars can intercept traffic without privileged init containers. Federated `ClusterSPIFFEID` resources ensure that workloads and gateways receive trust bundles from both clusters.
```

Reviewer 3: This short description is longer than the maximum 300 characters that will be allowed when we migrate to DITA.

.Prerequisites

* You have configured {SMProductName} for multi-cluster integration. For more information, see "Configuring {SMProductName} for multi-cluster {spire-full} integration" steps.
* The environment variables from the SPIRE deployment are set in your shell session.

> * You have installed {SMProductName} 3 on both clusters.

Reviewer 2: ```suggestion
* You have installed {SMProductName} {attribute} on both clusters.
```
version attribute instead of hard-coded 3?

.Procedure

. Create the {SMProductName} CNI namespace on Cluster A by running the following command:
+
[source,terminal]
----
$ oc new-project "${OSSM_CNI}" --kubeconfig="${CLUSTER_A_KUBECONFIG}" 2>/dev/null || true
----

> . Deploy the `IstioCNI` custom resource on Cluster A by running the following command:

Reviewer 2: ```suggestion
. Deploy the `IstioCNI` CR on Cluster A by running the following command:
```

+
[source,terminal,subs="+quotes,verbatim"]
----

> $ oc apply --kubeconfig="${CLUSTER_A_KUBECONFIG}" -f - <<EOF

Reviewer 2: separate yaml from command to apply the object

apiVersion: sailoperator.io/v1
kind: IstioCNI
metadata:
  name: default
spec:
  namespace: ${OSSM_CNI}
EOF
----

. Wait for the `IstioCNI` DaemonSet to become ready on Cluster A by running the following command:
+

> [source,terminal,subs="+quotes,verbatim"]

Reviewer 3: The following code block seems to have two commands in it. Per OCP doc guidelines, each code block should have only one command.

Reviewer 4: @wgabor0427 Are you accepting this comment?

----
$ until oc get daemonset/istio-cni-node --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n "${OSSM_CNI}" &> /dev/null; do
  sleep 3
done
$ oc rollout status daemonset/istio-cni-node --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n "${OSSM_CNI}" --timeout=300s
----

. Create the {SMProductName} CNI namespace on Cluster B by running the following command:
+
[source,terminal]
----
$ oc new-project "${OSSM_CNI}" --kubeconfig="${CLUSTER_B_KUBECONFIG}" 2>/dev/null || true
----

. Deploy the `IstioCNI` custom resource on Cluster B by running the following command:
+
[source,terminal,subs="+quotes,verbatim"]
----

> $ oc apply --kubeconfig="${CLUSTER_B_KUBECONFIG}" -f - <<EOF

Reviewer 2: (sep all; I won't comment on each one)

apiVersion: sailoperator.io/v1
kind: IstioCNI
metadata:
  name: default
spec:
  namespace: ${OSSM_CNI}
EOF
----

. Wait for the `IstioCNI` DaemonSet to become ready on Cluster B by running the following command:
+

> [source,terminal,subs="+quotes,verbatim"]

Reviewer 3: Similar comment here about multiple commands in one code block. Make sure each code block has only one command.

Reviewer 4: @wgabor0427 Are you accepting this comment?

----
$ until oc get daemonset/istio-cni-node --kubeconfig="${CLUSTER_B_KUBECONFIG}" -n "${OSSM_CNI}" &> /dev/null; do
  sleep 3
done
$ oc rollout status daemonset/istio-cni-node --kubeconfig="${CLUSTER_B_KUBECONFIG}" -n "${OSSM_CNI}" --timeout=300s
----

. Create the federated `ClusterSPIFFEID` resources on Cluster A by running the following command:
+

> [NOTE]

Reviewer 2: ```suggestion
```
this should be an IMPORTANT if doing this wrong breaks something
also, this should be below the steps, not between "do this" and the thing we need to do

> ====

Reviewer 2: ```suggestion
```

> Do not patch the default `ClusterSPIFFEID` (`zero-trust-workload-identity-manager-spire-default`). The {zero-trust-full} reconciles and reverts the manual changes. Instead, create custom `ClusterSPIFFEID` resources for the specific namespaces.

Reviewer 2: ```suggestion
```

> ====

Reviewer 2: ```suggestion
```

+
[source,terminal]
----

> $ oc apply --kubeconfig="${CLUSTER_A_KUBECONFIG}" -f - <<EOF

Reviewer 2: don't break these two up; this is clearly creating a single resource

apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterSPIFFEID
metadata:
  name: sample-federation
spec:
  className: zero-trust-workload-identity-manager-spire
  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
  namespaceSelector:
    matchLabels:
      kubernetes.io/metadata.name: sample
  federatesWith:
    - "${CLUSTER_B_TRUST_DOMAIN}"
---
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterSPIFFEID
metadata:
  name: istio-system-federation
spec:
  className: zero-trust-workload-identity-manager-spire
  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
  namespaceSelector:
    matchLabels:
      kubernetes.io/metadata.name: istio-system
  federatesWith:
    - "${CLUSTER_B_TRUST_DOMAIN}"
EOF
----

> 

Reviewer 2: ```suggestion
+
[IMPORTANT}
```
admonition goes here
